modules/core/src/main/scala/ch/epfl/scala/debugadapter/internal/evaluator/RuntimeEvaluator.scala

```diff
@@ -0,0 +1,475 @@
+      } yield result
+
+    /* -------------------------------------------------------------------------- */
+    /*                              Module evaluation                             */
+    /* -------------------------------------------------------------------------- */
+    @inline private def evaluateModule(tree: ModuleTree): Safe[JdiObject] =
+      Safe(JdiObject(tree.`type`.instances(1).get(0), frame.thread))
+
+    /* -------------------------------------------------------------------------- */
+    /*                                Instantiation                               */
+    /* -------------------------------------------------------------------------- */
+    @inline private def instantiate(tree: NewInstanceTree): Safe[JdiObject] =
+      for {
+        args <- tree.args.map(evaluate).traverse
+        boxedUnboxedArgs <- JdiObject.boxUnboxOnNeed(tree.method.argumentTypes(), args, frame)
+        instance <- JdiClass(tree.`type`, frame.thread).newInstance(tree.method, boxedUnboxedArgs)
+      } yield instance
+  }
+
+  /* -------------------------------------------------------------------------- */
+
+  private object RuntimeValidation {
+    case class Call(fun: Term, argClause: Term.ArgClause)
+    case class PreparedCall(qual: Validation[RuntimeValidationTree], name: String)
+
+    def validate(expression: Stat): Validation[RuntimeValidationTree] = {
+      expression match {
+        case value: Term.Name => validateName(value)
+        case _: Term.This => thisTree.map(tree => ThisTree(tree.value))
+        case sup: Term.Super => ???
+        case _: Term.Apply | _: Term.ApplyInfix | _: Term.ApplyUnary => validateMethod(extractCall(expression))
+        case select: Term.Select => validateSelect(select)
+        case newValue: Defn.Val => ???
+        case lit: Lit => Valid(LiteralTree(MatchingMethods.fromLitToValue(frame, lit)))
+        case instance: Term.New => validateNew(instance)
+      }
+    }
+
+    /* -------------------------------------------------------------------------- */
+    /*                               Name validation                              */
+    /* -------------------------------------------------------------------------- */
+    @inline private def extractElementIfReference(
+        tree: Validation[RuntimeValidationTree]
+    )(f: ReferenceType => Validation[RuntimeEvaluationTree]) =
+      tree match {
+        case ReferenceTree(ref) => f(ref)
+        case _: Invalid => staticCallError
+        case t => illegalAccess(t, "ReferenceType")
+      }
+
+    @inline private def varTreeByName(name: String): Validation[RuntimeEvaluationTree] =
+      Validation.fromOption(frame.variableByName(name)).map(LocalVarTree(_))
+
+    @inline private def fieldTreeByName(
+        of: Validation[RuntimeValidationTree],
+        name: String
+    ): Validation[RuntimeEvaluationTree] =
+      extractElementIfReference(of) { ref =>
```

Comment:
I tend to prefer to use `map` instead of custom higher order function

```suggestion
      asReference(of).map { ref =>
```